### ISSUE TITLE
Cannot set trainable parameters in PennyLane gradient transforms with Tensorflow

### ISSUE THREAD

**Issue by rmoyard** [PennyLaneAI/pennylane#3786]
### Feature details

Currently it is not possible to set trainable parameters in a PennyLane gradient transform with Tensorflow, because it has no requires grad and argnum is only a subset of trainable params.

### Implementation

We could use argnum to set the trainable parameters as well.

### How important would you say this feature is?

1: Not important. Would be nice to have.

**Comment by josh146** [comment]
> because it has no requires grad

I guess the TF equivalent of `requires_grad` is the `tape.watch()` method:

```pycon
>>> x = tf.constant([0.6, 0.1])
>>> with tf.GradientTape() as tape:
...     print(requires_grad(x))
False
>>> with tf.GradientTape() as tape:
...     tape.watch([x])
...     print(requires_grad(x))
True
```

The _default_ behaviour of `tape.watch()`, I believe, is that `tf.Variable` is watched, and `tf.constant` isn't (which makes sense!).

**Comment by josh146** [comment]
Oh actually @rmoyard I suppose my comment only makes sense if you have a `GradientTape` context. Are you thinking of use-cases where the user might not be using a TF gradient tape?

**Comment by rmoyard** [comment]
```
import tensorflow as tf
dev = qml.device("default.qubit", wires=2)


@qml.qnode(dev, interface="tf")
def circuit(weights):
    qml.RX(weights[0], wires=0)
    qml.RY(weights[1], wires=0)
    return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))


weights = tf.Variable([0.1, 0.2])

with tf.GradientTape() as tape:
    res = qml.gradients.param_shift(circuit)(weights)
```

Would you expect that to work?

**Comment by josh146** [comment]
Yes, I would expect `qml.math.requires_grad` _inside_ the `GradientTape()` context to return `True`. But _outside_ the gradient tape context it should return `False`. Is this not the case?

**Comment by josh146** [comment]
```python
weights = tf.Variable([0.1, 0.2])
qml.math.requires_grad(weights)  # False

with tf.GradientTape() as tape:
    res = qml.gradients.param_shift(circuit)(weights)
    qml.math.requires_grad(weights)  # True

qml.math.requires_grad(weights)  # False
```

**Comment by rmoyard** [comment]
Thanks @josh146, it works for trainability but the example is failing because of the transform itself unfortunately!
